sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

```diff
@@ -130,6 +130,15 @@ case class BatchScanExec(
         }
         k.copy(expressions = expressions, numPartitions = newPartValues.length,
           partitionValues = newPartValues)
+      case k: KeyGroupedPartitioning if spjParams.joinKeyPositions.isDefined =>
+        val expressions = spjParams.joinKeyPositions.get.map(i => k.expressions(i))
+        val newPartValues = k.partitionValues.map{r =>
```

Comment:
Just curious why this is necessary. I see that in BatchScanExec we are also doing something similar, by checking the `joinKeyPositions` and call `uniquePartitionValues` at the end to de-dup.

Reply:
Yea,  uniquePartitionValues is called at the end (the deferred grouping, while calculating the BatchScanExec's inputRDD), so its a bit too late.  This is doing the same in outputPartitioning (if its not the partiallyClusteredDistribution case), to make the two sides match during this scenario.